File: Solutions/Image_Overlap/Image_Overlap.go
func largestOverlap(A [][]int, B [][]int) int {
	aValues := make([]int, len(A))
	bValues := make([]int, len(B))
	for i, vs := range A {
		for j, v := range vs {
			if v == 1 {
				aValues[i] += 1 << j
			}
			if B[i][j] == 1 {
				bValues[i] += 1 << j
			}
		}
	}
	maxValue := 0
	// i: row offset
	// j: col offset
	for i := 0; i < len(A); i++ {
		for j := 0; j < len(A); j++ {
			current1, current2 := 0, 0
			for m := 0; m + i < len(A); m++ {
				current1 += getOne(aValues[m] & (bValues[i+m] >> j))
				current2 += getOne(bValues[m] & (aValues[i+m] >> j))
			}
			if current1 > maxValue {
				maxValue = current1
			}
			if current2 > maxValue {
				maxValue = current2
			}
		}
	}
	return maxValue
}

func getOne(v int) int {
	n := 0
	for v != 0 {
		if v & 0x1 == 1 {
			n++
		}
		v >>= 1
	}
	return n
}

//Fastest
func largestOverlap(A [][]int, B [][]int) int {
    var count int
    for i := -len(A)+1; i < len(A); i++ {
        for j := -len(A)+1; j < len(A);j++ {
            countMaybe := intersect(A, B, i, j)
            if count < countMaybe {
                count = countMaybe
            }
        }
    }
    return count
}

func intersect(A, B [][]int, offI, offJ int) int {
    var count int
    for i := range A {
        for j := range A[0] {
            if i + offI < 0 || j+offJ < 0 || i+offI>=len(A) || j+offJ>= len(A) {
                continue
            }
            if A[i+offI][j+offJ] + B[i][j] == 2  {
                count++
            }
        }
    }
    return count
}

File: Solutions/Image_Overlap/Image_Overlap.cpp
//Key idea:

//1. only considers the point with 1, so O(N1 * N2) where N1 and N2 are the number of points with 1 in each image
//2. Image A can be shifted horizontally in range [-M+1, M-1] and vertically in [-N+1, N-1], where + means right or down and - means left or up
//3. For each pair of points, update counter array of size 2M x 2N which keeps track of the number of overlapping points for the shift amount to overlap them where (M, N) is the origin.

class Solution {
public:
    int largestOverlap(vector<vector<int>>& A, vector<vector<int>>& B) {
        // vector of pair<r, c> for 1  
        int M = A.size(); // row count
        int N = A[0].size(); // column count 
        vector<pair<int, int>> oneA, oneB;
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (A[i][j])
                    oneA.emplace_back(i, j);
                if (B[i][j])
                    oneB.emplace_back(i, j);
            }
        }
        
        vector<vector<int>> C(2*M, vector<int>(2*N, 0));
        // for each pair, compute the diff in offset <row shift ro, column shift co>
        // for convenience, (M, N) is the origin
        for (auto a : oneA) {
            for (auto b : oneB) {
                C[b.first + M - a.first][b.second + N - a.second]++;
            }
        }
        
        int maxCnt = 0;
        for (int i = 0; i < 2*M; i++) {
            for (int j = 0; j < 2*N; j++) {
                if (C[i][j] > maxCnt)
                    maxCnt = C[i][j];
            }
        }
        // return max
        return maxCnt;
    }
};

//Fastest
class Solution {
public:
    int count(unsigned int num) {
      int cnt = 0;
      while (num) {
        cnt += (num & 1);
        num >>= 1;
      }
      return cnt;
    }

    int largestOverlap(vector<vector<int>>& A, vector<vector<int>>& B) {
      int largest = 0;
      int n = A.size();
      vector<unsigned int> arrA(n);
      vector<unsigned int> arrB(n);
      for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
          arrA[i] |= A[i][j] << (n - 1 - j);
          arrB[i] |= B[i][j] << (n - 1 - j);
        }
      }
      for (int s = 0; s < n; ++s) {
        for (int t = 0; t < n; ++t) {
          int cnt = 0;
          for (int i = s; i < n; ++i) {  /* A slides down s, right t */
            cnt += count((arrA[i - s] >> t) & arrB[i]);
          }
          largest = max(cnt, largest);
          cnt = 0;
          for (int i = 0; i < n - s; ++i) {  /* A slides up s, left t */
            cnt += count(arrB[i] & (arrA[i + s] << t));
          }
          largest = max(cnt, largest);
        }
      }
      return largest;
    }
};
